torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -324,127 +258,141 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
-            self.load_local_state_dict(ZeroRedundancyOptimizer.rank_local_state_dict(self.rank, state_dict))
+        # NOTE: In practice, `broadcast` is used, which is wasteful (gather would have been appropriate)
+        # compatibility issues with some backends make the use of broadcast mandatory for now.
+        # a possible follow up would be to move all sharded state management to RPC RRef
 
-    def local_state_dict(self) -> Dict:
-        """Gets this rank's ``state_dict``.
+        self._all_states = []
 
-        Returns:
-            The state of the optimizer as a :class:`dict`.
-            It contains two entries:
+        with torch.cuda.device(self._device) if self._device.type == torch.device(
+            "cuda"
+        ).type else contextlib.suppress():
+            empty_messenger = torch.tensor([0], dtype=torch.uint8, device=self._device)
+
+            for rank in range(self.world_size):
+                global_rank = _get_global_rank(self.group, rank)
+
+                # This rank collects the whole state
+                if self.rank == recipient_rank:
+                    if rank == self.rank:
+                        self._all_states.append(
+                            _recursive_copy_to_device(
+                                self.optim.state_dict(), non_blocking=True, device=torch.device("cpu")
+                            )
+                        )
+                    else:
+                        # Fetch the optim state from the other replicas
+                        replica_state = [empty_messenger]
+                        dist.broadcast_object_list(object_list=replica_state, src=global_rank, group=self.group)
 
-            * state - a dict holding current optimization state. Its content
-                differs between optimizer classes.
-            * param_groups - a dict containing all parameter groups
-        """
-        return self.optim.state_dict()
+                        self._all_states.append(
+                            _recursive_copy_to_device(replica_state[0], non_blocking=True, device=torch.device("cpu"))
+                        )
+                else:
+                    # Acknowledge broadcasts, and send this rank's shard when needed
+                    # Default to CPU space to gain some memory headroom
+                    if rank == self.rank:
+                        # Send the state to the reference replica
+                        dist.broadcast_object_list(
+                            object_list=[self.optim.state_dict()], src=self.global_rank, group=self.group
+                        )
 
-    def state_dict(self) -> Dict[str, Any]:
+                    elif rank != recipient_rank:
+                        # Discard this tensor/rank, broadcast was being use for compatibility reasons
+                        dist.broadcast_object_list(object_list=[empty_messenger], src=global_rank, group=self.group)
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
```

Comment:
so, this is another big change, since we can now load a "normal" state dict. The index gymnastic is mostly symmetrical to saving, except that we're only interested in updating the state for this rank's optimizer shard, the rest is not used